network_wrangler/utils/transit.py

```diff
@@ -0,0 +1,3087 @@
+
+    # Make roadway network nodes a GeoDataFrame if it's not already
+    if not isinstance(roadway_net.nodes_df, gpd.GeoDataFrame):
+        # Convert to GeoDataFrame if needed
+        roadway_net.nodes_df = gpd.GeoDataFrame(
+            roadway_net.nodes_df,
+            geometry=gpd.points_from_xy(roadway_net.nodes_df.X, roadway_net.nodes_df.Y),
+            crs=LAT_LON_CRS,
+        )
+
+    # Collect bus stops to match
+    feed_tables["stops"]["is_bus_stop"] = False
+    feed_tables["stops"].loc[
+        feed_tables["stops"]["route_types"].apply(
+            lambda x: RouteType.BUS in x if isinstance(x, list) else False
+        )
+        | feed_tables["stops"]["route_types"].apply(
+            lambda x: RouteType.TROLLEYBUS in x if isinstance(x, list) else False
+        ),
+        "is_bus_stop",
+    ] = True
+    WranglerLogger.debug(f"feed_tables['stops'].value_counts():\n{feed_tables['stops']}")
+    bus_stops_gdf = (
+        feed_tables["stops"].loc[feed_tables["stops"]["is_bus_stop"] == True].copy(deep=True)
+    )
+
+    # Reset index to ensure continuous indices from 0 to n-1
+    bus_stops_gdf.reset_index(drop=True, inplace=True)
+
+    WranglerLogger.info(f"Matching {len(bus_stops_gdf):,} bus stops to roadway nodes")
+    WranglerLogger.debug(f"bus_stops_gdf:\n{bus_stops_gdf}")
+    WranglerLogger.debug(f"feed_tables['shapes']:\n{feed_tables['shapes']}")
+    if trace_shape_ids:
+        for trace_shape_id in trace_shape_ids:
+            WranglerLogger.debug(
+                f"trace shapes for {trace_shape_id}:\n"
+                f"{feed_tables['shapes'].loc[feed_tables['shapes']['shape_id'] == trace_shape_id]}"
+            )
+
+    # Build BallTree for bus-access nodes that are in the bus graph
+    # This ensures we only match to connected nodes
+    G_bus = roadway_net.get_modal_graph("bus")
+    bus_graph_nodes = set(G_bus.nodes())
+    roadway_net.nodes_df["bus_access"] = False
+    roadway_net.nodes_df.loc[
+        roadway_net.nodes_df["model_node_id"].isin(bus_graph_nodes), "bus_access"
+    ] = True
+    WranglerLogger.debug(
+        f"bus-accessible nodes from graph: {roadway_net.nodes_df['bus_access'].sum():,}"
+    )
+
+    # Extract bus-accessible nodes
+    bus_accessible_nodes_gdf = roadway_net.nodes_df[
+        roadway_net.nodes_df["bus_access"] == True
+    ].copy()
+    WranglerLogger.debug(
+        f"Found {len(bus_accessible_nodes_gdf):,} bus-accessible nodes (for mixed-traffic transit) "
+        f"out of {len(roadway_net.nodes_df):,} total"
+    )
+
+    # Project bus nodes and bus_stops_gdf to specified CRS and bus_stops
+    bus_accessible_nodes_gdf.to_crs(local_crs, inplace=True)
+    bus_stops_gdf.to_crs(local_crs, inplace=True)
+
+    # Initialize results
+    bus_stops_gdf["model_node_id"] = None
+    bus_stops_gdf[f"match_distance_{crs_units}"] = np.inf  # in crs_units
+
+    # Build spatial index for bus nodes
+    try:
+        from sklearn.neighbors import BallTree  # noqa: PLC0415
```

Comment:
Nice! I bet this is a lot faster than the default RTREE that geopandas uses.  Maybe we should create a `netpandas` extension